ops/testing.py

```diff
@@ -1518,14 +1518,63 @@ def add_layer(
                 if service.override not in ('merge', 'replace'):
                     raise RuntimeError('500 Internal Server Error: layer "{}" has invalid '
                                        '"override" value on service "{}"'.format(label, name))
-                if service.override != 'replace':
+                if service.override == 'replace':
+                    layer.services[name] = service
+                elif service.override == 'merge':
+                    if combine:
+                        self._merge_service(label, name, service)
+                    else:
+                        layer.service[name] = service
+                else:
                     raise RuntimeError(
                         'override: "{}" unsupported for layer "{}" service "{}"'.format(
                             service.override, label, name))
-                layer.services[name] = service
+
         else:
             self._layers[label] = layer_obj
 
+    def _merge_service(self, label, name: str, service: pebble.Service):
+        layer = self._layers[label]
+        if name not in layer.services:
+            layer.service[name] = service
+            return
+
+        s = layer.services[name]
+        if service.summary:
+            s.summary = service.summary
+        if service.description:
+            s.description = service.description
+        if service.startup:
+            s.startup = service.startup
+        if service.command:
+            s.command = service.command
+        if service.user_id:
+            s.user_id = service.user
+        if service.user:
+            s.user = service.user
+        if service.group_id:
+            s.group_id = service.group_id
+        if service.group:
+            s.group = service.group
+        s.after = s.after + service.after
+        s.before = s.before + service.before
+        s.requires = s.requires + service.requires
+        for key in service.environment:
+            s.environment[key] = service.environment[key]
+        if service.on_success:
+            s.on_success = service.on_success
+        if service.on_failure:
+            s.on_failure = service.on_failure
+        for key in service.on_check_failure:
+            s.on_check_failure[key] = service.on_check_failure[key]
+        if service.backoff_delay:
+            s.backoff_delay = service.backoff_delay
+        if service.backoff_limit:
+            s.backoff_limit = service.backoff_limit
+        if service.backoff_factor:
+            s.backoff_factor = service.backoff_factor
+        layer.services[name] = s
```

Comment:
This approach works, but seems fragile.

In general, objects in the ops framework have a lot of properties, for easy reference, but also maintain an internal dict of values. You can access that dictionary for the Service object by calling .to_dict()

So you could do something like:

```
for key in service.to_dict():
    # Setup an anonymous function that will update the service definition.
    setter = lambda: setattr(s, key, getattr(service, key))

    # Handle special cases
    if key  == 'requires':
         setter = setattr(s, key, s.requires + service.requires)
    if key == 'environment':
        setter = lambda: None
        for key in service.environment:
            s.environment[key] = service.environment[key]
    # ... etc.

    # Call the anonymous function (it will pick up the correct values from its context)
    setter()
```

I think that's cleaner, and it will automagically pick up any new keys that don't need special handling on merge. The code as written is very clear, limitations and all, though, so feel free to push back on that :-)

Reply:
If this is just a mirror of how we do it in pebble, I think that's fine.

Maybe we should add a comment to the Pebble code, noting that the Ops harness needs to be updated if the merge function is updated. (We could even add pebble as a dependency to the testing harness, though I'm not 100% happy with coupling the libraries too tightly together.)

Reply:
also that single-char variable should make my eyes bleed, so I'm told by many reviewers of my own PRs

Reply:
well.. pebble code is written in Go.. i think the approach suggested by @pengale is ok, so that's what i did.. i anybody strongly objects i can return it to the way it was before (i.e. as it in pebble)

Reply:
I love single char identifiers!  Kids these days :P